src/fdcache_fdinfo.cpp

```diff
@@ -225,25 +356,695 @@ bool PseudoFdInfo::ClearUntreated(off_t start, off_t size)
     return untreated_list.ClearParts(start, size);
 }
 
-bool PseudoFdInfo::GetUntreated(off_t& start, off_t& size, off_t max_size, off_t min_size)
+bool PseudoFdInfo::GetLastUntreated(off_t& start, off_t& size, off_t max_size, off_t min_size)
 {
     AutoLock auto_lock(&upload_list_lock);
 
-    return untreated_list.GetPart(start, size, max_size, min_size);
+    return untreated_list.GetLastUpdatedPart(start, size, max_size, min_size);
 }
 
-bool PseudoFdInfo::GetLastUntreated(off_t& start, off_t& size, off_t max_size, off_t min_size)
+bool PseudoFdInfo::AddUntreated(off_t start, off_t size)
 {
     AutoLock auto_lock(&upload_list_lock);
 
-    return untreated_list.GetLastUpdatedPart(start, size, max_size, min_size);
+    bool result = untreated_list.AddPart(start, size);
+    if(!result){
+        S3FS_PRN_DBG("Failed adding untreated area part.");
+    }else if(S3fsLog::IsS3fsLogDbg()){
+        untreated_list.Dump();
+    }
+
+    return result;
 }
 
-bool PseudoFdInfo::AddUntreated(off_t start, off_t size)
+bool PseudoFdInfo::GetLastUpdateUntreatedPart(off_t& start, off_t& size)
+{
+    // Get last untreated area
+    if(!untreated_list.GetLastUpdatePart(start, size)){
+        return false;
+    }
+    return true;
+}
+
+bool PseudoFdInfo::ReplaceLastUpdateUntreatedPart(off_t front_start, off_t front_size, off_t behind_start, off_t behind_size)
+{
+    if(0 < front_size){
+        if(!untreated_list.ReplaceLastUpdatePart(front_start, front_size)){
+            return false;
+        }
+    }else{
+        if(!untreated_list.RemoveLastUpdatePart()){
+            return false;
+        }
+    }
+    if(0 < behind_size){
+        if(!untreated_list.AddPart(behind_start, behind_size)){
+            return false;
+        }
+    }
+    return true;
+}
+
+//
+// Utility for sorting upload list
+//
+static bool filepart_partnum_compare(const filepart& src1, const filepart& src2)
+{
+    return (src1.get_part_number() <= src2.get_part_number());
+}
+
+bool PseudoFdInfo::InsertUploadPart(off_t start, off_t size, int part_num, bool is_copy, etagpair** ppetag, bool lock_already_held)
+{
+    //S3FS_PRN_DBG("[start=%lld][size=%lld][part_num=%d][is_copy=%s]", static_cast<long long int>(start), static_cast<long long int>(size), part_num, (is_copy ? "true" : "false"));
+
+    if(!IsUploading()){
+        S3FS_PRN_ERR("Multipart Upload has not started yet.");
+        return false;
+    }
+    if(start < 0 || size <= 0 || part_num < 0 || !ppetag){
+        S3FS_PRN_ERR("Parameters are wrong.");
+        return false;
+    }
+
+    AutoLock auto_lock(&upload_list_lock, lock_already_held ? AutoLock::ALREADY_LOCKED : AutoLock::NONE);
```

Comment:
Now this method is private and when calling this method the upload_list_lock is already locked then.
However, since this method call may be added in the future, I made a lock processing to operate upload_list.